% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/sigfit_utility.R
\name{build_catalogues}
\alias{build_catalogues}
\title{Build mutational catalogues}
\usage{
build_catalogues(variants)
}
\arguments{
\item{variants}{Character matrix with one row per single-nucleotide variant, and four (or five) columns:
\itemize{
 \item{Sample ID (character, e.g. "Sample 1").}
 \item{Reference allele (character: "A", "C", "G", or "T").}
 \item{Alternate allele (character: "A", "C", "G", or "T").}
 \item{Trinucleotide context of the variant (character; the reference sequence between the positions 
 immediately before and after the variant; e.g. "TCA").}
 \item{Optional: transcriptional strand of the variant (character: "T" for transcribed,
 or "U" for untranscribed). If this column is included, a strandwise representation of
 catalogues will be used.}
}}
}
\value{
An integer matrix of mutation counts, where each row corresponds to a sample and each column
corresponds to one of the 96 (or 192) trinucleotide mutation types.
}
\description{
\code{build_catalogues} generates a set of mutational catalogues from a table containing 
the base change and trinucleotide context of each single-nucleotide variant in every sample.
}
\examples{
# Load example mutation data
data("variants_21breast")
head(variants_21breast)

# Build catalogues
counts <- build_catalogues(variants_21breast)
counts
}
