#' Single base substitution signatures from COSMIC v3.2
#'
#' A data set of 78 SBS signatures from COSMIC v3.2 (March 2021).
#' With a few exceptions, these signatures were extracted
#' using SigProfiler (as described in Alexandrov et al., 2020) from the 2,780
#' whole-genome variant calls produced by the ICGC/TCGA Pan Cancer Analysis of
#' Whole Genomes (PCAWG) Network.
#' The following signatures are annotated as "Possible sequencing artefacts":
#' SBS27, SBS43, SBS45-60.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/SBS/}
#' @source \url{https://cancer.sanger.ac.uk/signatures/downloads/}
"cosmic_signatures_v3.2"

#' Small insertion/deletion signatures from COSMIC v3.2
#'
#' A data set of 18 ID signatures from COSMIC v3.2 (March 2021).
#' With a few exceptions, these signatures were extracted
#' using SigProfiler (as described in Alexandrov et al., 2020) from the 2,780
#' whole-genome variant calls produced by the ICGC/TCGA Pan Cancer Analysis of
#' Whole Genomes (PCAWG) Network.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/ID/}
#' @source \url{https://cancer.sanger.ac.uk/signatures/downloads/}
"cosmic_signatures_indel_v3.2"

#' Doublet base substitution signatures from COSMIC v3.2
#'
#' A data set of 11 DBS signatures from COSMIC v3.2 (March 2021).
#' With a few exceptions, these signatures were extracted
#' using SigProfiler (as described in Alexandrov et al., 2020) from the 2,780
#' whole-genome variant calls produced by the ICGC/TCGA Pan Cancer Analysis of
#' Whole Genomes (PCAWG) Network.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/ID/}
#' @source \url{https://cancer.sanger.ac.uk/signatures/downloads/}
"cosmic_signatures_doublet_v3.2"

#' Single base substitution signatures from COSMIC v3
#'
#' A data set of 67 SBS signatures from COSMIC v3.0 (May 2019).
#' These signatures were extracted using SigProfiler (as described in Alexandrov
#' et al., 2020) from the 2,780 whole-genome variant calls produced by the
#' ICGC/TCGA Pan Cancer Analysis of Whole Genomes (PCAWG) Network.
#' The following signatures are annotated as "Possible sequencing artefacts":
#' SBS27, SBS43, SBS45-60.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/SBS/}
#' @source \url{https://www.synapse.org/#!Synapse:syn11801889}
"cosmic_signatures_v3"

#' Single base substitution signatures from COSMIC v3 with transcriptional strand bias
#'
#' A data set of 65 transcriptional strand-wise SBS signatures from COSMIC v3.0 (May 2019).
#' These signatures were extracted using SigProfiler (as described in Alexandrov
#' et al., 2020) from the 2,780 whole-genome variant calls produced by the
#' ICGC/TCGA Pan Cancer Analysis of Whole Genomes (PCAWG) Network.
#' The following signatures are annotated as "Possible sequencing artefacts":
#' SBS27, SBS43, SBS45-60.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/SBS/}
#' @source \url{https://www.synapse.org/#!Synapse:syn11801889}
"cosmic_signatures_strand_v3"

#' Single base substitution signatures from COSMIC v2
#'
#' A data set of 30 SBS signatures from COSMIC v2 (March 2015).
#' These signatures were extracted using the WTSI Mutational Signature Framework
#' as described in Alexandrov et al., 2013.
#'
#' @source \url{https://cancer.sanger.ac.uk/cosmic/signatures/SBS/}
#' @source \url{https://cancer.sanger.ac.uk/signatures/downloads/}
"cosmic_signatures_v2"
